plugins/org.jkiss.dbeaver.ui.editors.sql.ai/src/org/jkiss/dbeaver/ui/editors/sql/ai/controls/ScopeConfigDialog.java

```diff
@@ -0,0 +1,208 @@
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.swt.widgets.Tree;
+import org.eclipse.swt.widgets.TreeItem;
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.DBException;
+import org.jkiss.dbeaver.Log;
+import org.jkiss.dbeaver.model.DBIcon;
+import org.jkiss.dbeaver.model.DBPDataSource;
+import org.jkiss.dbeaver.model.DBUtils;
+import org.jkiss.dbeaver.model.navigator.DBNDatabaseNode;
+import org.jkiss.dbeaver.model.navigator.DBNUtils;
+import org.jkiss.dbeaver.model.runtime.AbstractJob;
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+import org.jkiss.dbeaver.model.struct.DBSEntity;
+import org.jkiss.dbeaver.model.struct.DBSObject;
+import org.jkiss.dbeaver.model.struct.DBSObjectContainer;
+import org.jkiss.dbeaver.model.struct.DBSStructContainer;
+import org.jkiss.dbeaver.ui.DBeaverIcons;
+import org.jkiss.dbeaver.ui.UIUtils;
+import org.jkiss.dbeaver.ui.dialogs.BaseDialog;
+import org.jkiss.dbeaver.utils.GeneralUtils;
+import org.jkiss.utils.ArrayUtils;
+
+import java.util.*;
+
+public class ScopeConfigDialog extends BaseDialog {
+    private static final Log log = Log.getLog(ScopeConfigDialog.class);
+
+    private Tree objectTree;
+    private final Set<String> checkedObjectIds;
+    private final DBPDataSource dataSource;
+
+    public ScopeConfigDialog(@NotNull Shell shell, @NotNull Set<String> checkedIds, @NotNull DBPDataSource dataSource) {
+        super(shell, "Customize scope", DBIcon.AI);
+        this.checkedObjectIds = new LinkedHashSet<>(checkedIds);
+        this.dataSource = dataSource;
+    }
+
+    @Override
+    protected Composite createDialogArea(Composite parent) {
+        Composite composite = super.createDialogArea(parent);
+
+        objectTree = new Tree(composite, SWT.CHECK);
+        GridData gd = new GridData(GridData.FILL_BOTH);
+        gd.widthHint = 400;
+        gd.heightHint = 300;
+        objectTree.setLayoutData(gd);
+        objectTree.addSelectionListener(new SelectionAdapter() {
+            @Override
+            public void widgetSelected(SelectionEvent e) {
+                if (e.detail != SWT.CHECK) {
+                    return;
+                }
+                TreeItem item = (TreeItem) e.item;
+                if (item.getData() instanceof DBSStructContainer) {
+                    checkTreeItems(item.getItems(), item.getChecked());
+                }
+            }
+
+            private void checkTreeItems(TreeItem[] items, boolean check) {
+                for (TreeItem child : items) {
+                    child.setChecked(check);
+                    if (child.getData() instanceof DBSEntity) {
+
+                    } else {
+                        TreeItem[] children = child.getItems();
+                        if (!ArrayUtils.isEmpty(children)) {
+                            checkTreeItems(children, check);
+                        }
+                    }
```

Comment:
Isn't it better to invert condition and move the code from else to if?

Reply:
It's extracted code, but okay.